### ISSUE TITLE
Scripting: cannot cast from [def] to [void] as consumer

### ISSUE THREAD

**Issue by stu-elastic** [elastic/elasticsearch#92821]
When someone uses the fields api to `emit` to a runtime field such as `field('foo').forEach(this::emit)`, painless throws a class cast exception "cannot cast from [def] to [void]" due to the branch
```
// for method invocation, its allowed to ignore the return value
if (interfaceMethod.returnType() == void.class) {
      returnType = def.class;
```

in [DefaultSemanticAnalysisPhase.visitLambda](https://github.com/elastic/elasticsearch/blob/53e500db539a52bb714c9ec7dd27de52032ba69a/modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultSemanticAnalysisPhase.java#L2341).

Updating that line leads to a few failing unit tests, [`LambdaTests.testVoidReturn`](https://github.com/elastic/elasticsearch/blob/8913b7141c8d8b28f3b50cbaf7d1161ea9874750/modules/lang-painless/src/test/java/org/elasticsearch/painless/LambdaTests.java#L93) and [`UserFunctionTests.testCallUserMethodFromStatementWithinNestedLambda`](https://github.com/elastic/elasticsearch/blob/57ac810951a4309bfddddcc9892531366d8769c3/modules/lang-painless/src/test/java/org/elasticsearch/painless/UserFunctionTests.java#L181)

It looks like we may need a couple of changes to fix this issue.

A) Add a decoration when a consumer is expected and, if the lambda returns a value, add a pop instruction to swallow the return value.
B) For the `def` case, in `LambdaBootstrap`, add a proxy where `def` is able to cast to `void` and the value is swallowed.  We may be able to do this with an update methodhandle.

**Comment by thecoop** [comment]
@stu-elastic is there a standalone test case that can reproduce this in a unit test?

**Comment by stu-elastic** [comment]
Not yet, feel free to take a pass if ya have time.

**Comment by thecoop** [comment]
Yeah, I've been trying to come up with a test case and failing, so far

**Comment by stu-elastic** [comment]
It's difficult, I can repro against server but can't get a test case after a few obvious attempts.

**Comment by jdconrad** [comment]
Closing this as it is unlikely to be prioritized.
